### DISCUSSION TITLE
[@types/multer] My nestjs web app requires @types/multer package in production

### DISCUSSION THREAD

**Discussion by ngohuytrieu** [DefinitelyTyped/DefinitelyTyped#69721] [Issues with a @types package]
Hi guy, I am deploying my nestjs web app to production and have the following issue
cc @jt000, @DavidBR-SW, @mxl, @HyunSeob, @PierreTchuente, @thrymgjol, @peterblazejewicz

I upload file using multer packages:
```bash
npm i -S multer
npm i -D @types/multer
```
I write a nest js pipe to validate my uploading file
```typescript
@Injectable()
export class FileValidationPipe implements PipeTransform {
  constructor(private readonly files?: string[]) {}

  async transform(value: Express.Multer.File): Promise<object> {}
}
```

It works fine in development environment, but when deploying to production environment I encountered this issue 
```bash
transform(value: Express.Multer.File)
Cannot find namespace 'Express'
```

If I move `@types/multer` package to `dependencies` instead of `devDependencies` it works ok but I don't want that because I only need it in development

**Version:**
Node: `v18.17.0`
Multer: `^1.4.5-lts.1`
@types/multer: `^1.4.10`

[a comment by an automated account is not shown]

**Comment by peterblazejewicz** [comment]
Hey, looks like issue with your build pipeline setup, not with types. 'dev' is for dev-time, including build pipeline dependencies. So your build pipeline seems to be run with PROD environment setup (or a parameter for NPM script run), hence types are not installed. Though it means your 'typescript' is dependency, as 'tsc' (compiler) runs and fails to find types.
in the essence, not a problem with types itself. Thx!

**Comment by ngohuytrieu** [comment]
@peterblazejewicz so if I run a compile command in PROD environment, I have to install all required @types packages for the pipeline?

**Comment by peterblazejewicz** [comment]
that's not related to environment/pipeline. Some would argue `-D` is proper target for installation (development only), e.g. if needed only to run say ESLint against your source code.  `tsc` compiler requires types/*** to be resolvable at build/compile time. Exact answer depends on your build setup. We are not answering CI/CD answers here.

**Comment by Tommy1223454** [comment]
If you already have a NestJS pipe for upload validation, I’d think of that as a great place to separate “basic request acceptance” from “do we trust this file enough to keep/process it”.

In production, I’ve found it useful to make that second step explicit and content-aware instead of relying only on metadata or filename-level validation.

I built an OSS package for this in Node.js:
https://github.com/pompelmi/pompelmi

It’s meant for inspect-before-storage workflows around untrusted uploads.
